app/scripts/modules/core/src/search/global/GlobalSearch.tsx

```diff
@@ -0,0 +1,409 @@
+    const refToFocus = flattenedRefs[flattenedRefs.length - 1];
+    refToFocus && refToFocus.focus();
+  }
+
+  private queryChanged({ target }: React.ChangeEvent<HTMLInputElement>) {
+    this.setState({ query: target.value }, () => {
+      if (this.state.query.length >= 3) {
+        this.executeQuery();
+      } else {
+        this.considerMinLengthWarning();
+      }
+    });
+  }
+
+  private navigateResult(event: React.KeyboardEvent<HTMLAnchorElement>) {
+    const { which, target } = event;
+    if (which === 27) { // escape
+      ReactGA.event({ category: 'Global Search', action: 'Keyboard Nav', label: 'escape (from result)' });
+      this.reset();
+    }
+    if (which === 9) { // tab - let it navigate automatically, but close menu if on the last result
+      const flattenedRefs = flatten(this.resultRefs);
+      const lastResultRef = flattenedRefs[flattenedRefs.length - 1];
+      if (target === lastResultRef) {
+        ReactGA.event({ category: 'Global Search', action: 'Keyboard Nav', label: 'tab (from result)' });
+        this.hideResults();
+        return;
+      }
+    }
+    if (which === 40) { // down
+      ReactGA.event({ category: 'Global Search', action: 'Keyboard Nav', label: 'down (from result)' });
+      const flattenedRefs = flatten(this.resultRefs);
+      const currentRefIndex = flattenedRefs.indexOf((target as HTMLElement));
+      const nextResultRef = flattenedRefs[currentRefIndex + 1];
+
+      nextResultRef && nextResultRef.focus();
+      event.preventDefault();
+    }
+    if (which === 38) { // up
+      ReactGA.event({ category: 'Global Search', action: 'Keyboard Nav', label: 'up (from result)' });
+      const flattenedRefs = flatten(this.resultRefs);
+      const currentRefIndex = flattenedRefs.indexOf((target as HTMLElement));
+      const prevResultRef = flattenedRefs[currentRefIndex - 1];
+
+      prevResultRef && prevResultRef.focus();
+      event.preventDefault();
+    }
+  }
+
+  @Debounce(300)
+  private executeQuery() {
+    const { infrastructureSearchService } = ReactInjector;
+    const search = infrastructureSearchService.getSearcher();
+
+    ReactGA.event({ category: 'Global Search', action: 'Query', label: this.state.query });
+    this.setState({ querying: true, showMinLengthWarning: false });
+
+    search.query(this.state.query).then((result) => {
+      const showMinLengthWarning = !!this.state.query && this.state.query.length < 3;
+
+      const categories = result
+        .filter(({ results }) => results.length)
+        .map((category) => ({
+          ...category,
+          results: searchRank(category.results).slice(0, 5)
+        }))
+        .sort((a, b) => a.type.order - b.type.order);
+
+      this.resultRefs = range(categories.length).map(() => []);
+
+      this.setState({
+        querying: false,
+        showSearchResults: !!this.state.query && !showMinLengthWarning,
+        categories,
+        showMinLengthWarning
+      });
+    });
+  }
+
+  @Debounce(300)
```

Comment:
why debounce this?

Reply:
oh my original comment on this line got lost from a force push :)

I originally had this showing/hiding based on immediately updated logic in `render` which is a lot simpler, but the UX of seeing a red warning pop up as you're typing is *really* jarring and awkward if you're not actually typing a search that's too short. Its a little annoying to deal this as an async state change, but I'm ok with the tradeoff. Do you think it's worth throwing a comment in here for posterity?

Reply:
ah, gotcha.  
I guess a comment couldn't hurt?